qa/tasks/cephfs/test_volumes.py

```diff
@@ -5295,6 +5295,234 @@ def test_clone_subvolume_info(self):
+        self.assertEqual(len(result['pending_clones']), 1)
+
+        # check clone status
+        self._wait_for_clone_to_complete(clone, clone_group=target_group)
+
+        # remove snapshot
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot, group)
+
+        # remove subvolumes
+        self._fs_cmd("subvolume", "rm", self.volname, subvolume, group)
+        self._fs_cmd("subvolume", "rm", self.volname, clone, target_group)
+
+        # remove groups
+        self._fs_cmd("subvolumegroup", "rm", self.volname, group)
+        self._fs_cmd("subvolumegroup", "rm", self.volname, target_group)
+
+        # verify trash dir is clean
+        self._wait_for_trash_empty()
+
+    def test_subvolume_snapshot_info_if_orphan_clone(self):
+        """
+        Verify subvolume snapshot info output if orphan clones exists.
+        Orphan clones should not list under pending clones.
+        orphan_clones_count should display correct count of orphan clones'
+        """
+        subvolume = self._generate_random_subvolume_name()
+        snapshot = self._generate_random_snapshot_name()
+        clone_list =  [f'clone_{i}' for i in range(3)]
+
+        # create subvolume.
+        self._fs_cmd("subvolume", "create", self.volname, subvolume, "--mode=777")
+
+        # snapshot subvolume
+        self._fs_cmd("subvolume", "snapshot", "create", self.volname, subvolume, snapshot)
+
+        # insert delay at the beginning of snapshot clone
+        self.config_set('mgr', 'mgr/volumes/snapshot_clone_delay', 5)
+
+        # schedule a clones
+        for clone in clone_list:
+            self._fs_cmd("subvolume", "snapshot", "clone", self.volname, subvolume, snapshot, clone)
+
+        # remove track file for third clone to make it orphan
+        meta_path = os.path.join(".", "volumes", "_nogroup", subvolume, ".meta")
+        pending_clones_result = self.mount_a.run_shell(f"sudo grep \"clone snaps\" -A3 {meta_path}", omit_sudo=False, stdout=StringIO(), stderr=StringIO())
+        third_clone_track_id = pending_clones_result.stdout.getvalue().splitlines()[3].split(" = ")[0]
+        third_clone_track_path = os.path.join(".", "volumes", "_index", "clone", third_clone_track_id)
+        self.mount_a.run_shell(f"sudo rm -f {third_clone_track_path}", omit_sudo=False)
+
+        # list snapshot info
+        result = json.loads(self._fs_cmd("subvolume", "snapshot", "info", self.volname, subvolume, snapshot))
+
+        # verify snapshot info
+        expected_clone_list = []
+        for i in range(len(clone_list)-1):
+            expected_clone_list.append({"name": clone_list[i]})
+        self.assertEqual(result['has_pending_clones'], "yes")
+        self.assertEqual(result['orphan_clones_count'], 1)
+        self.assertListEqual(result['pending_clones'], expected_clone_list)
+        self.assertEqual(len(result['pending_clones']), 2)
+
+        # check clone status
+        self._wait_for_clone_to_complete(clone)
```

Comment:
@nmshelke: This fails here: https://pulpito.ceph.com/vshankar-2022-06-27_06:36:09-fs-wip-vshankar-testing-20220627-100931-testing-default-smithi/6900237/

```
022-06-27T09:18:25.746 INFO:tasks.cephfs_test_runner:======================================================================
2022-06-27T09:18:25.747 INFO:tasks.cephfs_test_runner:FAIL: test_subvolume_snapshot_info_if_orphan_clone (tasks.cephfs.test_volumes.TestSubvolumeSnapshotClones)
2022-06-27T09:18:25.747 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2022-06-27T09:18:25.748 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2022-06-27T09:18:25.748 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5292901bc999daa0d7ac128764a3f27c4233c252/qa/tasks/cephfs/test_volumes.py", line 5511, in test_subvolume_snapshot_info_if_orphan_clone
2022-06-27T09:18:25.748 INFO:tasks.cephfs_test_runner:    self._wait_for_clone_to_complete(clone)
2022-06-27T09:18:25.749 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5292901bc999daa0d7ac128764a3f27c4233c252/qa/tasks/cephfs/test_volumes.py", line 66, in _wait_for_clone_to_complete
2022-06-27T09:18:25.749 INFO:tasks.cephfs_test_runner:    self.__check_clone_state("complete", clone, clone_group, timo)
2022-06-27T09:18:25.749 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5292901bc999daa0d7ac128764a3f27c4233c252/qa/tasks/cephfs/test_volumes.py", line 55, in __check_clone_state
2022-06-27T09:18:25.750 INFO:tasks.cephfs_test_runner:    self.assertTrue(check < timo)
2022-06-27T09:18:25.750 INFO:tasks.cephfs_test_runner:AssertionError: False is not true
2022-06-27T09:18:25.750 INFO:tasks.cephfs_test_runner:
2022-06-27T09:18:25.751 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
```

Doesn't seem related to this change as such as it is waiting for clone to finish under a specific time. Could you take a look?

Reply:
@vshankar This was failing in teuthology because test was waiting to complete clone for orphaned clone (which is not possible).

Fixed the issue.

Please find test results below:
http://pulpito.front.sepia.ceph.com/nshelke-2022-06-30_09:38:16-fs:volumes-wip-vshankar-testing-20220627-100931-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/nshelke-2022-06-30_11:06:41-fs:volumes-wip-vshankar-testing-20220627-100931-distro-default-smithi/